src/main/java/com/jcraft/jsch/OpenSshCertificate.java

```diff
@@ -0,0 +1,317 @@
+  private final long validAfter;
+
+  // match ssh-keygen behavior where the default would be forever
+  private final long validBefore;
+
+  /**
+   * Critical options that must be recognized by the SSH implementation
+   */
+  private final Map<String, String> criticalOptions;
+
+  /**
+   * Extensions that provide additional functionality
+   */
+  private final Map<String, String> extensions;
+
+  /**
+   * Reserved field for future use
+   */
+  private final String reserved;
+
+  /**
+   * The CA's public key that signed this certificate
+   */
+  private final byte[] signatureKey;
+
+  /**
+   * The cryptographic signature of the certificate
+   */
+  private final byte[] signature;
+
+  /**
+   * The certificate data without the certificate type and the signature
+   */
+  private final byte[] message;
+
+  /**
+   * Private constructor to be used exclusively by the Builder.
+   */
+  private OpenSshCertificate(Builder builder) {
+    this.keyType = builder.keyType;
+    this.nonce = builder.nonce;
+    this.certificatePublicKey = builder.certificatePublicKey;
+    this.serial = builder.serial;
+    this.type = builder.type;
+    this.id = builder.id;
+    this.principals = builder.principals;
+    this.validAfter = builder.validAfter;
+    this.validBefore = builder.validBefore;
+    this.criticalOptions = builder.criticalOptions;
+    this.extensions = builder.extensions;
+    this.reserved = builder.reserved;
+    this.signatureKey = builder.signatureKey;
+    this.signature = builder.signature;
+    this.message = builder.message;
+  }
+
+  public String getKeyType() {
+    return keyType;
+  }
+
+  public byte[] getNonce() {
+    return nonce;
+  }
+
+  public byte[] getCertificatePublicKey() {
+    return certificatePublicKey;
+  }
+
+  public long getSerial() {
+    return serial;
+  }
+
+  public int getType() {
+    return type;
+  }
+
+  public String getId() {
+    return id;
+  }
+
+  public Collection<String> getPrincipals() {
+    return principals;
+  }
+
+  public long getValidAfter() {
+    return validAfter;
+  }
+
+  public long getValidBefore() {
+    return validBefore;
+  }
+
+  public Map<String, String> getCriticalOptions() {
+    return criticalOptions;
+  }
+
+  public Map<String, String> getExtensions() {
+    return extensions;
+  }
+
+  public String getReserved() {
+    return reserved;
+  }
+
+  public byte[] getSignatureKey() {
+    return signatureKey;
+  }
+
+  public byte[] getSignature() {
+    return signature;
+  }
+
+  public boolean isUserCertificate() {
+    return SSH2_CERT_TYPE_USER == type;
+  }
+
+  public boolean isHostCertificate() {
+    return SSH2_CERT_TYPE_HOST == type;
+  }
+
+  public boolean isValidNow() {
+    return OpenSshCertificateUtil.isValidNow(this);
+  }
+
+  public byte[] getMessage() {
+    return message;
+  }
+
+  /**
+   * A static inner builder class for creating immutable OpenSshCertificate instances.
+   */
+  public static class Builder {
```

Comment:
Let's make this class package private (along with all fields & methods that are public within it).